## Basic A/B Testing using a simulated data

In [3]:
import pandas as pd
import numpy as np

np.random.seed(123)

n = 1000

group = ["A"] * n + ["B"] * n
conversion = list(np.random.binomial(1, 0.10, n)) + \
             list(np.random.binomial(1, 0.12, n))

df = pd.DataFrame({
    "group": group,
    "conversion": conversion
})

df

,group,conversion
0,A,0
1,A,0
2,A,0
3,A,0
4,A,0
...,...,...
1995,B,0
1996,B,0
1997,B,0
1998,B,0


In [5]:
df['conversion'].value_counts()

conversion
0    1767
1     233
Name: count, dtype: int64

In [6]:
# Conversion rates
df.groupby("group")["conversion"].mean()

group
A    0.104
B    0.129
Name: conversion, dtype: float64

## Statistical test

In [7]:
from statsmodels.stats.proportion import proportions_ztest

success = df.groupby("group")["conversion"].sum().values
nobs = df.groupby("group")["conversion"].count().values

stat, pval = proportions_ztest(success, nobs)

print("Z-stat:", stat)
print("P-value:", pval)

Z-stat: -1.7424437670240522
P-value: 0.08143082312036741


The p-value of 0.081 indicates that the result is not statistically significant at the 5% level, so we fail to reject the null hypothesis of no effect. However, the negative Z-statistic suggests the treatment may reduce conversion. Given that the p-value is close to 0.10, I would describe this as weak evidence of a negative effect rather than no effect. I would next examine the effect size, confidence intervals, and statistical power before making a product decision.